### PR TITLE
User can store customer data

### PR DESCRIPTION
**Pull request by sundqvistjohan**
[PT Link](https://www.pivotaltracker.com/story/show/170302017)
Adds scenarios for adding a new customer including sad path for incorrect form input
Adds customer controller and model
Adds "Add new customer" link to "Invoices index/dashboard" leading to form for adding a customer. Customers are rendered on "Invoices index".

### CODE REVIEW

File: app/controllers/customers_controller.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  end
+
+  def create
+    customer = Customer.create(customer_params)
+
+    if customer.persisted?
+      redirect_to invoices_path, notice: 'Customer was succesfully added.'
+    else
+      redirect_to request.referer, notice: 'Please fill in the form correctly.'
+      end
```
Comment (oliverochman) [inline app/controllers/customers_controller.rb]: indentation

File: app/controllers/customers_controller.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  end
+
```
Comment (oliverochman) [inline app/controllers/customers_controller.rb:19]: u need to add private here

File: app/controllers/customers_controller.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+class CustomersController < ApplicationController
```
Comment (oliverochman) [inline app/controllers/customers_controller.rb:3]: can a visitor create a customer?

**Review by oliverochman — CHANGES_REQUESTED** [review]
Will jump in with you guys and set up the association between the user and customer

File: app/controllers/customers_controller.rb (hunk 3 above)
Comment (sundqvistjohan) [reply app/controllers/customers_controller.rb]: absolutely not. fixed.

File: app/controllers/customers_controller.rb (hunk 1 above)
Comment (sundqvistjohan) [reply app/controllers/customers_controller.rb]: fixed

File: app/controllers/customers_controller.rb (hunk 2 above)
Comment (sundqvistjohan) [reply app/controllers/customers_controller.rb:19]: done

File: app/controllers/customers_controller.rb (hunk 3 above)
Comment (sundqvistjohan) [reply app/controllers/customers_controller.rb:3]: no way jose, added authentication check

**Comment by sundqvistjohan** [comment]
adds reference between Customer and User and only shows current user his own customers

File: features/user_can_store_customer_data.feature [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  In order to avoid manual input for repeated invoices
+  I would like to store my customers information
+
+  Background:
+    Given the following users exist
+      | email          | password |
+      | user@user.com  | password |
+      | user2@user.com | password |
+
+    Given the following customers exist
```
Comment (oliverochman) [inline features/user_can_store_customer_data.feature]: Change this to `And`

File: features/user_can_store_customer_data.feature [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    And I fill in "Address" with "Valhallavägen 79"
+    And I fill in "Postcode" with "string"
+    And I fill in "City" with "Stockholm"
+    And I click on "Add customer"
+    Then I should see "Please fill in the form correctly."
+
+  Scenario: User can see customers created by itself
+    Given I am logged in as "user@user.com"
+    And I am on the index page
+    Then I should see "Company"
```
Comment (oliverochman) [inline features/user_can_store_customer_data.feature]: Please make this assertion a bit clearer. Maybe go for another company name.

File: features/user_can_store_customer_data.feature [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    And I should see "123456-7890"
+
+  Scenario: User can not see customers created by another user
+    Given I am logged in as "user2@user.com"
+    And I am on the index page
+    Then I should not see "Company"
```
Comment (oliverochman) [inline features/user_can_store_customer_data.feature]: Same as on line 45

**Review by oliverochman — CHANGES_REQUESTED** [review]

File: features/user_can_store_customer_data.feature (hunk 5 above)
Comment (sundqvistjohan) [reply features/user_can_store_customer_data.feature]: Yeah good call to not have the company name the same as the label...

**Review by oliverochman — APPROVED** [review]

**Pull request merged**